omnigan/trainer.py

```diff
@@ -1351,106 +926,42 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
                 if self.opts.dis.p.use_local_discriminator:
                     fake_d_global = self.D["p"]["global"](fake)
                     real_d_global = self.D["p"]["global"](x)
+
                     fake_d_local = self.D["p"]["local"](fake * m)
                     real_d_local = self.D["p"]["local"](x * m)
 
-                    global_loss = self.losses["D"]["p"](
-                        fake_d_global, False, True
-                    ) + self.losses["D"]["p"](real_d_global, True, True)
+                    global_loss = self.losses["D"]["p"](fake_d_global, False, True)
+                    global_loss += self.losses["D"]["p"](real_d_global, True, True)
 
-                    local_loss = self.losses["D"]["p"](
-                        fake_d_local, False, True
-                    ) + self.losses["D"]["p"](real_d_local, True, True)
+                    local_loss = self.losses["D"]["p"](fake_d_local, False, True)
+                    local_loss += self.losses["D"]["p"](real_d_local, True, True)
 
                     disc_loss["p"]["global"] += global_loss
                     disc_loss["p"]["local"] += local_loss
                 else:
-                    real_fake_d = self.D["p"](
-                        torch.cat(
-                            [torch.cat([m, x], axis=1), torch.cat([m, fake], axis=1)],
-                            axis=0,
-                        )
-                    )
-                    fake_d, real_d = divide_pred(real_fake_d)
+                    real_cat = torch.cat([m, x], axis=1)
+                    fake_cat = torch.cat([m, fake], axis=1)
+                    real_fake_cat = torch.cat([real_cat, fake_cat], dim=0)
+                    real_fake_d = self.D["p"](real_fake_cat)
```

Comment:
This is a big change. It was a bug. I don't understand how the painter can be so good with this shit lying around. It's not updating the discriminator in the  right direction!

Reply:
@sashavor @melisandeteng

Reply:
I may be blind but I don't see the difference with the previous version ?

Reply:
ouuuuh I see it! wow weird

Reply:
Maybe divide_pred should be modified to return "real, fake" instead of "fake, real" then, so it's clearer? Although it's just names.

Reply:
divide_pred splits tensors it does not care about order, it all depends on the order of concatenation

Reply:
but then let's not call them "fake,real" in the return of the function?:)

Reply:
I agree, it's just the way it is in the SPADE code, not sure it's worth changing :p

Reply:
I just updated it actually to follow your recommendations